File: Data Structures & Algorithms/target-sum/submission-2.java

class Solution {
    Map<String, Integer> dp = new HashMap<>();

    public int findTargetSumWays(int[] nums, int target) {
        return find(nums, target, 0, 0);
    }

    int find(int[] nums, int target, int cur, int i) {
        if (i == nums.length)
            return cur == target ? 1 : 0;

        String key = i + "," + cur;

        if (dp.containsKey(key))
            return dp.get(key);

        int ans = find(nums, target, cur + nums[i], i + 1)
                + find(nums, target, cur - nums[i], i + 1);

        dp.put(key, ans);

        return ans;
    }
}

File: Data Structures & Algorithms/combination-sum-iv/submission-2.py
class Solution:
    def combinationSum4(self, nums: List[int], target: int) -> int:
        dp = {}

        def find(t):
            if t == 0:
                return 1

            if t < 0:
                return 0
            if(t in dp):
                return dp[t]
            ans = 0

            for num in nums:
                ans += find(t - num)

            dp[t] = ans
            return ans
        return find(target)